src/LibraryManager/Manifest.cs

```diff
@@ -251,48 +256,143 @@ public void Uninstall(string libraryId, Action<string> deleteFileAction)
         /// that deletes the files from the project.
         /// </remarks>
         /// <param name="deleteFileAction">>An action to delete the files.</param>
+        /// <param name="cancellationToken"></param>
         /// <returns></returns>
-        public IEnumerable<ILibraryInstallationResult> Clean(Action<string> deleteFileAction)
+        public async Task<IEnumerable<ILibraryInstallationResult>> CleanAsync(Func<IEnumerable<string>, Task<bool>> deleteFileAction, CancellationToken cancellationToken)
         {
-            List<ILibraryInstallationResult> results = new List<ILibraryInstallationResult>();
+            var results = new List<ILibraryInstallationResult>();
 
             foreach (ILibraryInstallationState state in Libraries)
             {
-                results.Add(DeleteLibraryFiles(state, deleteFileAction));
+                results.Add(await DeleteLibraryFilesAsync(state, deleteFileAction, cancellationToken));
             }
 
             return results;
         }
 
-        private ILibraryInstallationResult DeleteLibraryFiles(ILibraryInstallationState state, Action<string> deleteFileAction)
+        /// <summary>
+        /// Removes unwanted library files
+        /// </summary>
+        /// <param name="newManifest"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<bool> RemoveUnwantedFilesAsync(Manifest newManifest, CancellationToken cancellationToken)
         {
-            int filesDeleted = 0;
+            if (newManifest != null)
+            {
+                IEnumerable<FileIdentifier> existingFiles = await GetAllManifestFilesWithVersionsAsync(Libraries).ConfigureAwait(false);
+                IEnumerable<FileIdentifier> newFiles = await GetAllManifestFilesWithVersionsAsync(newManifest.Libraries).ConfigureAwait(false);
+                IEnumerable<string> filesToRemove = existingFiles.Where(f => !newFiles.Contains(f)).Select(f => f.Path);
+
+                if (filesToRemove.Any())
+                {
+                    IHostInteraction hostInteraction = _dependencies.GetHostInteractions();
+                    return await hostInteraction.DeleteFilesAsync(filesToRemove, cancellationToken);
+                }
+            }
 
-            IProvider provider = _dependencies.GetProvider(state.ProviderId);
-            ILibraryInstallationResult updatedStateResult = provider.UpdateStateAsync(state, CancellationToken.None).Result;
+            return false;
+        }
 
-            if (updatedStateResult.Success)
+        private async Task<IEnumerable<FileIdentifier>> GetAllManifestFilesWithVersionsAsync(IEnumerable<ILibraryInstallationState> libraries)
+        {
+            var files = new List<FileIdentifier>();
```

Comment:
Use a HashSet to make the `Contains` call faster?  

Does order matter in the return value?